src/mame/heathkit/tlb.h

```diff
@@ -0,0 +1,126 @@
+// license:BSD-3-Clause
+// copyright-holders:Mark Garlanger
+/***************************************************************************
+
+    Heathkit Terminal Logic Board (TLB)
+
+****************************************************************************/
+
+#ifndef MAME_HEATHKIT_TLB_H
+#define MAME_HEATHKIT_TLB_H
+
+#pragma once
+
+#include "cpu/z80/z80.h"
+#include "machine/ins8250.h"
+#include "machine/mm5740.h"
+#include "sound/beep.h"
+#include "video/mc6845.h"
+#include "emupal.h"
+#include "screen.h"
+#include "speaker.h"
+
+class heath_tlb_device : public device_t
+{
+public:
+	heath_tlb_device(const machine_config &mconfig, const char *tag, device_t *owner, uint32_t clock);
+
+	// interface routines
+	auto serial_data_callback() { return m_write_sd.bind(); }
+
+	DECLARE_WRITE_LINE_MEMBER(cb1_w);
+
+protected:
+	heath_tlb_device(const machine_config &mconfig, device_type type, const char *tag, device_t *owner, uint32_t clock);
+
+	virtual ioport_constructor device_input_ports() const override;
+	virtual const tiny_rom_entry *device_rom_region() const override;
+	virtual void device_start() override;
+	virtual void device_add_mconfig(machine_config &config) override;
+	virtual void device_resolve_objects() override;
+
+private:
+
+	void mem_map(address_map &map);
+	void io_map(address_map &map);
+
+	void key_click_w(uint8_t data);
+	void bell_w(uint8_t data);
+	uint8_t kbd_key_r();
+	uint8_t kbd_flags_r();
+	uint16_t translate_mm5740_b(uint16_t b);
+
+	void checkBeepState();
+
+	void serial_out_b(uint8_t data);
+
+	DECLARE_READ_LINE_MEMBER(mm5740_shift_r);
+	DECLARE_READ_LINE_MEMBER(mm5740_control_r);
+	DECLARE_WRITE_LINE_MEMBER(mm5740_data_ready_w);
+
+	MC6845_UPDATE_ROW(crtc_update_row);
+
+	TIMER_CALLBACK_MEMBER(key_click_off);
+	TIMER_CALLBACK_MEMBER(bell_off);
+
+	emu_timer *m_key_click_timer = nullptr;
+	emu_timer *m_bell_timer = nullptr;
+
+	devcb_write_line m_write_sd;
+
+	required_device<palette_device> m_palette;
+	required_device<cpu_device>     m_maincpu;
+	required_device<mc6845_device>  m_crtc;
+	required_device<ins8250_device> m_ace;
+	required_device<beep_device>    m_beep;
+	required_shared_ptr<uint8_t>    m_p_videoram;
+	required_region_ptr<u8>         m_p_chargen;
+	required_device<mm5740_device>  m_mm5740;
+	required_memory_region          m_kbdrom;
+	required_ioport                 m_kbspecial;
+
+	uint8_t  m_transchar;
+	bool     m_strobe;
+	bool     m_keyclickactive;
+	bool     m_bellactive;
+
+};
+
+#define USE_ALT_BIOS 0
+#if !USE_ALT_BIOS
```

Comment:
It can’t go in like this – you need to decide which way it’s going to work and go with that.

Reply:
Agreed, I should have been clearer here, I only added it here, so you could see the code I added to support the alternative ROMs. I wanted to see if there was a different way which doesn't require all ROMs to be available. I think I will use the the classes, we can revisit this if the internal UI can be improved to not require all the ROMs. I'm not familiar with that code, but it seems like order of the checks could be changed. It could store which of the ROMs are available, but continue to then read the config, and only then if the needed ROMs are not available should it error.

Reply:
MAME full sets have been available on the Internet Archive for 5+ years now, so there really is no reason to be timid about requiring ROMs.